common/src/main/cpp/MotionProfiling.cpp

```diff
@@ -1,76 +1,71 @@
+  _ntbounds.emplace_back(_table, "kD", &_kD);
+  _ntbounds.emplace_back(_table, "kV", &_kV);
+  _ntbounds.emplace_back(_table, "kA", &_kA);
+  _ntbounds.emplace_back(_table, "kG", &_kG);
+}
+
+double PathfinderGains::GetkP() const {
+  return _kP;
+}
+
+double PathfinderGains::GetkI() const {
+  return _kI;
+}
+
+double PathfinderGains::GetkD() const {
+  return _kD;
+}
+
+double PathfinderGains::GetkV() const {
+  return _kV;
 }
 
-// Pathfinder MP
+double PathfinderGains::GetkA() const {
+  return _kA;
+}
 
-void PathfinderMPMode::init() {
-  _ecfg = {
-    _enc->GetEncoderTicks(),
-    _enc->GetEncoderTicksPerRotation(), _cfg.wheel_diameter * 0.0254 * PI, _cfg.kp, _cfg.ki, _cfg.kd, _cfg.kv, _cfg.ka
-  };
-  _follow = { 0, 0, 0, 0, 0 };
+double PathfinderGains::GetkG() const {
+  return _kG;
 }
 
-double PathfinderMPMode::calculate() {
-  double ret = pf_follow_enc(_ecfg, &_follow, _segments, _seg_length, _enc->GetEncoderRotations());
-  done = _follow.finished;
-  return ret;
+PathfinderController::PathfinderController(PathfinderGains gains) : _gains(gains) {
+  Reset();
 }
 
-// void TalonMPMode::init() {
-//   _talon->reset_mp();
-//   _talon->SetSelectedSensorPosition(0, 0, 0);
-//   _talon->configure_mp_update_rate(_segments[0].dt * 1000);
-//   _talon->configure_encoder_edges_per_rev(_cfg.enc_ticks_per_rev);
-//   _talon->configure_wheel_diameter(_cfg.wheel_diameter);
-//   _talon->configure_pidf(
-//     _cfg.kp, _cfg.ki, _cfg.kd,
-//     1023.0 / _cfg.kv 
-//   );
-//   _talon->load_pathfinder(_segments, _seg_length);
-//   _talon->Set(ctre::phoenix::motion::SetValueMotionProfile::Disable);
-// }
-
-// // Talon FP
-
-// double TalonMPMode::calculate() {
-//   _talon->load_pathfinder(_segments, _seg_length);
-//   auto status = _talon->process_mp();
-
-//   done = status.isLast;
-//   return status.isLast ? ctre::phoenix::motion::SetValueMotionProfile::Hold : ctre::phoenix::motion::SetValueMotionProfile::Enable;
-// }
+void PathfinderController::Load(std::string project, std::string pathname) {
+  Reset();
+  CurtinPathfinder::LoadDeployedFile(project, pathname + ".left", _segmentsL);
+  _trajLen = CurtinPathfinder::LoadDeployedFile(project, pathname + ".right", _segmentsR);
+}
+
+void PathfinderController::Reset() {
+  _cfg = FollowerConfig{ _gains.GetkP(), _gains.GetkI(), _gains.GetkD(), _gains.GetkV(), _gains.GetkA() };
+  _followerL = DistanceFollower{0, 0, 0, 0, false};
+  _followerR = DistanceFollower{0, 0, 0, 0, false};
+}
+
+void PathfinderController::SetOffset(double distanceL, double distanceR) {
+  _offsetL = distanceL;
+  _offsetR = distanceR;
+} 
+
+std::pair<double, double> PathfinderController::Calculate(double distL, double distR, double gyroAngle) {
+  double l = pathfinder_follow_distance(_cfg, &_followerL, _segmentsL, _trajLen, distL - _offsetL);
+  double r = pathfinder_follow_distance(_cfg, &_followerR, _segmentsR, _trajLen, distR - _offsetR);
+
+  double desiredAngle = -r2d(_followerL.heading);
```

Comment:
Just me getting triggered a little bit (I know it doesn't actually change anything as each side is the same), but on this line:
(https://github.com/CurtinFRC/2019-DeepSpace/blob/b54ec61469c6dd2bf8754d8779c61d82bd0f3803/common/src/main/cpp/MotionProfiling.cpp#L49)
you get the generic trajectory length from the right hand side, but here you get the bearing from the left hand side...

Reply:
The heading and file lengths are guaranteed to be identical, so it's not really a big problem